File: backend/src/models/ConsentRecord.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface IConsentRecord extends Document {
  patient_id: mongoose.Types.ObjectId;
  consent_type: string;
  granted_to: mongoose.Types.ObjectId;
  status: 'active' | 'revoked' | 'expired';
  granted_at: Date;
  expires_at?: Date;
  created_at: Date;
  updated_at: Date;
}

const ConsentRecordSchema = new Schema<IConsentRecord>(
  {
    patient_id: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    consent_type: { type: String, required: true },
    granted_to: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    status: { type: String, enum: ['active', 'revoked', 'expired'], default: 'active' },
    granted_at: { type: Date, default: Date.now },
    expires_at: { type: Date },
  },
  { timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' } }
);

export default mongoose.model<IConsentRecord>('ConsentRecord', ConsentRecordSchema);

File: backend-py/app/models/consent_record.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import Column, String, DateTime, Enum, ForeignKey
from sqlalchemy.dialects.postgresql import UUID

from app.database import Base


class ConsentRecord(Base):
    __tablename__ = "consent_records"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    patient_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    consent_type = Column(String, nullable=False)
    granted_to = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    status = Column(
        Enum("active", "revoked", "expired", name="consent_status_enum"),
        nullable=False,
        server_default="active",
    )
    granted_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    expires_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    def to_dict(self):
        return {
            "id": str(self.id),
            "_id": str(self.id),
            "patient_id": str(self.patient_id),
            "consent_type": self.consent_type,
            "granted_to": str(self.granted_to),
            "status": self.status,
            "granted_at": self.granted_at.isoformat() if self.granted_at else None,
            "expires_at": self.expires_at.isoformat() if self.expires_at else None,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }
